doc/design/tensor_design.md

```diff
@@ -0,0 +1,173 @@
+# Tensor Design
+
+
+## Allocation
+
+`Allocation` is a [RAII](http://en.cppreference.com/w/cpp/language/raii) class template, which is used to handle a piece of memory. 
+
+```cpp
+template <typename Device>
+class Allocation {
+ public:
+  Allocation();
+  Allocation(size_t size, Device device);
+
+  // Creates a non-owned allocation
+  Allocation(void *ptr, size_t size, Device device);
```

Comment:
Is there any other concept other than Tensor that will use `Allocation`? 
I think if we have not clearly used examples, we can remove the interface first. Such as the comment https://github.com/PaddlePaddle/Paddle/pull/2587#discussion_r123888812.